Frameworks/Foundation/NSURL.mm

```diff
@@ -890,4 +908,63 @@ - (BOOL)getPromisedItemResourceValue:(id _Nullable*)value forKey:(NSString*)key
     return StubReturn();
 }
 
+/**
+ @Status Interoperable
+*/
++ (instancetype)URLWithStorageFile:(RTObject<WSIStorageFile>*)storageFile {
+    return [[[self alloc] initWithStorageFile:storageFile] autorelease];
+}
+
+/**
+ @Status Interoperable
+*/
+- (instancetype)initWithStorageFile:(RTObject<WSIStorageFile>*)storageFile {
+    ComPtr<IInspectable> comPtr = [storageFile comObj];
+    ComPtr<IStorageItem> comStorageItem;
+    RETURN_NULL_IF_FAILED(comPtr.As(&comStorageItem));
+
+    // Now cache the storage item so that it can later be retreived.
+    ComPtr<IStorageApplicationPermissionsStatics> storageApplicationPermissionsStatics;
+    RETURN_NULL_IF_FAILED(
+        GetActivationFactory(Wrappers::HStringReference(RuntimeClass_Windows_Storage_AccessCache_StorageApplicationPermissions).Get(),
+                             &storageApplicationPermissionsStatics));
+
+    ComPtr<IStorageItemAccessList> accessList;
+    RETURN_NULL_IF_FAILED(storageApplicationPermissionsStatics->get_FutureAccessList(&accessList));
+
+    // See how close we are to the cap.
+    uint32_t maxItems = 0;
+    RETURN_NULL_IF_FAILED(accessList->get_MaximumItemsAllowed(&maxItems));
+
+    // Documentation says 1000 but lets just make sure it was somthing valid.
+    RETURN_NULL_IF_FALSE(maxItems > 0);
```

Comment:
>ULL_IF_FALSE(maxItems > 0); [](start = 12, length = 27)

Is there a future world where they're unlimited and managed by the system? We may not want to break on those platforms.